### PR TITLE
[pt2] add meta function for `linalg_qr`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100714

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -390,6 +390,59 @@ def linalg_inv_ex_meta(A: Tensor, check_errors: bool = False):
...
+        reduced = True  # this is actually irrelevant in this mode
+    else:
+        check(
+            False,
+            lambda: (
+                f"qr received unrecognized mode '{mode}' "
+                f"but expected one of 'reduced' (default), 'r', or 'complete'"
+            ),
+        )
+    return compute_q, reduced
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:412]: See `_parse_qr_mode` in core.

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -390,6 +390,59 @@ def linalg_inv_ex_meta(A: Tensor, check_errors: bool = False):
...
+        Q.as_strided_(Q_shape, make_contiguous_strides_for(Q_shape, row_major=False))
+    else:
+        Q = A.new_empty([0])
+
+    # For readability
+    R_shape = list(A.shape)
+    R_shape[-2] = k if reduced_mode or not compute_q else m
+    R = A.new_empty(R_shape)
+    R.as_strided_(R_shape, make_contiguous_strides_for(R_shape, row_major=False))
+    return Q, R
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:443]: See `TORCH_META_FUNC(linalg_qr)` in core.

**Review by ezyang — APPROVED** [review]

**Review by lezcano — APPROVED** [review]

**Pull request closed without merging**